Bugzilla integration

fgyanz opened the pull request:
Add new functionality to scan command. By default, with no options set, it will fetch all the list of pending
CVEs from bugzilla and analyse each one of them. Once done, it will print a report.

```
     ID  CVE         SUBSYSTEM       CVSS    PRIORITY      CLASSIFICATION    STATUS         AFFECTED
-------  ----------  ------------  ------   ----------   ----------------  -------------  --------------------
122xxxx  2021-xxxxx  scsi             7.8    Medium         complex          Incomplete(0)  No
122xxxx  2024-xxxxx  drm/amdgpu       7.8    Medium         trivial          Fixed(3)       15.6rtu0 15.6u0
122xxxx  2024-xxxxx  iommu/vt-d       7.8    High           None             Fixed(1)       15.6rtu0 15.6u0-1
```

It takes around 3-4min to scan 90 CVEs.

fgyanz committed:
klplib/bugzilla: Add more helper functions

Introduce some functions for better handling the information
fetched from bugzilla. Also rework and simplify the already
existing ones.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
tests: Fix and add bugzilla tests

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/ksrc: Fetch tags in a thread-safe way

git-fetch is not thread-safe by nature, so add a lock
to protect in case of concurrent access.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/kernel_tree: Fetch tags in a thread-safe way

git-fetch is not thread-safe by nature, so add a lock
to protect in case of concurrent access.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
plugins/scan: Integrate bugzilla into scan

- Adapt scan() to be called from multiple threads
at the same.

- Add new functionality to scan command. By default, with
no options set, it will fetch all the list of pending
CVEs from bugzilla and analyse each one of them. Once done,
it will print a report.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/ksrc: return upstream commits separately

Return upstream commits information in a separate variable for
easier access by the callers.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
man: Add bulk analysis to scan

Explain the bugzilla integration into scan and how
to interprete the reported information.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
README: Remove reference to klp-build script

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

marcosps reviewed (approved):
I liked it! Thanks a lot for working on it! Now that we depend on tabulate, we can start using it more widely to make the output of extract more easily to digest, and maybe setup as well...

vmezzela commented on klpbuild/klplib/bugzilla.py:
```diff
@@ -5,26 +5,172 @@
+
+def get_bug_data(bug, dep):
+    return (get_bug_cve(bug), get_bug_subsys(bug), get_bug_cvss(dep),
+            get_bug_classification(dep), get_bug_prio(bug))
```
```suggestion
            get_bug_classification(bug), get_bug_prio(bug))
```
Shouldn't we read the classification from the livepatching bug?

fgyanz replied:
Indeed, good catch. Thanks!

vmezzela commented on klpbuild/plugins/scan.py:
```diff
@@ -6,20 +6,24 @@
+import concurrent.futures
+import tabulate
 
+from tabulate import tabulate
```
redundant whit the previous line?

vmezzela commented on klpbuild/plugins/scan.py:
```diff
@@ -34,18 +38,80 @@ def register_argparser(subparser):
-    upstream = patches.get("upstream", [])
+    upstream, patches = get_patches(cve, savedir)
 
+    # Always get the latest supported.csv file and check the content
```
this comment has been dropped by `f4035ac`, I guess it came from a rebase :)

fgyanz replied:
Right!

vmezzela commented on klpbuild/plugins/scan.py:
```diff
@@ -34,17 +38,80 @@ def register_argparser(subparser):
+                                          "CLASSIFICATION", "STATUS", "AFFECTED"]))
+
+
+def scan_job(bug, dep, cve):
```
just noticed, we already have `get_bug_dep()`, which is what I meant by `bug.get_parent()`

fgyanz commented:
@marcosps yes, I do. Though I haven't done it yet because I've been working on another PR related to this one, and a few things might need to be changed here. But if I'm blocking you then I'll do one last push and merge. Sorry for the inconvenience!

fgyanz committed:
klplib:ksrc: Improve is_module_supported regex

Handle correctly corner cases of module names contained
within other names. For instance, when searching for:
'net/netfilter/ipset/ip_set_bitmap_ip'
The current regex also returns back:
'net/netfilter/ipset/ip_set_bitmap_ipmac'

Additionaly, check for supported markers like "+base".

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

marcosps commented:
> @marcosps yes, I do. Though I haven't done it yet because I've been working on another PR related to this one, and a few things might need to be changed here. But if I'm blocking you then I'll do one last push and merge. Sorry for the inconvenience!

not really a big deal, thanks!

fgyanz committed:
klplib/supported: Make cache R/W thread-safe

Access the cached list in a thread-safe way, so that
only one thread writes the cache once.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz commented on klpbuild/klplib/ksrc.py:
```diff
@@ -207,7 +212,7 @@ def get_patches(cve, savedir=None):
         logging.info("")
 
-    logging.info(f"upstream")
+    logging.info("upsteam")
```
Done!

fgyanz commented on klpbuild/klplib/ksrc.py:
```diff
@@ -43,10 +43,12 @@
-    This decorator checks whether the kernel-source tags are fetched. If not,
-    it fetches them the configuration and then calls the wrapped function.
+    This decorator checks whether the kernel-source tags are fetched in a
+    thread-safe way. If not, it fetches them the configuration and then calls
```
Done!

fgyanz commented on klpbuild/plugins/scan.py:
```diff
@@ -34,17 +38,80 @@ def register_argparser(subparser):
 
-
 def run(cve, conf, lp_filter, download):
+    no_check = False
```
Indeed, my bad!

fgyanz commented on klpbuild/plugins/scan.py:
```diff
@@ -6,20 +6,24 @@
 from klpbuild.klplib.supported import get_supported_codestreams
 from klpbuild.klplib.data import download_missing_cs_data
 from klpbuild.klplib.ksrc import get_patches, get_patched_kernels, cs_is_affected
+from klpbuild.klplib.bugzilla import get_pending_bugs, get_bug_data, is_bug_dropped, get_bug_dep, get_bug_summary
```
Done!

fgyanz commented on klpbuild/plugins/scan.py:
```diff
@@ -34,18 +38,80 @@ def register_argparser(subparser):
+    affected = "No"
+    status = "Not-Fixed"
+
+    patches, upstream, patched_cs, affected_cs = scan(cve, None, None, False)
```
Right, leftovers from a previous version of the code. Thanks!

fgyanz commented on klpbuild/klplib/bugzilla.py:
```diff
@@ -5,26 +5,172 @@
+    deps_fields = ["status", "resolution", "assigned_to", "whiteboard"]
+    deps = {d.id:d for d in __bzapi.getbugs(deps_ids,include_fields=deps_fields)}
+
+    return bugs, deps
```
Done!

fgyanz commented:
Hi guys! I pushed a newer version attending the comments and with some minor improvements. Thanks for reviewing the PR!

marcosps reviewed (approved):
LGTM! thanks a lot for working on it!

vmezzela reviewed (changes requested):
I love it! Thanks a lot @fgyanz, I left just one small request inline but besides that I'd merge the PR.
